src/model/warning.ts

```diff
@@ -150,3 +164,197 @@ const defaultPrinterOptions = {
   verbosity: 'full' as 'full',
   color: true
 };
+
+
+export type Action = EditAction | {
+  /** To ensure that type safe code actually checks for the action kind. */
+  kind: 'dummy-action-kind';
+};
+
+/**
+ * An EditAction is like a fix, only it's not applied automatically when the
+ * user runs `polymer lint --fix`. Often this is because it's less safe to
+ * apply automatically, and there may be caveats, or multiple ways to resolve
+ * the warning.
+ *
+ * For example, a change to an element that updates it to no longer use a
+ * deprecated feature, but that involves a change in the element's API should
+ * not be a fix, but should instead be an EditAction.
+ */
+export interface EditAction {
+  kind: 'edit';
+  /**
+   * A unique string code for the edit action. Useful so that the user can
+   * request that all actions with a given code should be applied.
+   */
+  code: string;
+  /**
+   * A short description of the change, noting caveats and important information
+   * for the user.
+   */
+  description: string;
+  edit: Edit;
+}
+
+/**
+ * Represents an action for replacing a range in a document with some text.
+ *
+ * This is sufficient to represent all operations on text files, including
+ * inserting and deleting text (using empty ranges or empty replacement
+ * text, respectively).
+ */
+export interface Replacement {
+  readonly range: SourceRange;
+  readonly replacementText: string;
+}
+
+
+/**
+ * A set of replacements that must all be applied as a single atomic unit.
+ */
+export type Edit = ReadonlyArray<Replacement>;
+
+export interface EditResult {
+  /** The edits that had no conflicts, and are reflected in editedFiles. */
+  appliedEdits: Edit[];
+
+  /** Edits that could not be applied due to overlapping ranges. */
+  incompatibleEdits: Edit[];
+
+  /** A map from urls to their new contents. */
+  editedFiles: Map<string, string>;
+}
+
+/**
+ * Takes the given edits and, provided there are no overlaps, applies them to
+ * the contents loadable from the given loader.
+ *
+ * If there are overlapping edits, then edits earlier in the array get priority
+ * over later ones.
+ */
+export async function applyEdits(
```

Comment:
why does the analyzer have code to apply edits?

Reply:
It's useful for multiple downstream projects. The CLI and the IDE.

I suppose both depend on the linter as well as the analyzer, but I feel like the applyEdits function is closely tied to the meaning of an Edit as an atomic unit of change.

Also the only legible way to test edits is to apply them and assert on the result. We'll probably want to add some fixes and edits in the core analyzer scanners, and we'll want to test them too.

Reply:
![OK](https://media.giphy.com/media/146heXDX89mUgw/giphy.gif)